docs/architecture/adr-057-RPC.md

```diff
@@ -0,0 +1,91 @@
+
+There are currently two options being discussed: gRPC & JSON-RPC.
+
+### JSON-RPC:
+
+JSON-RPC is a JSON-based RPC protocol. Tendermint has implemented its own variant of JSON-RPC which is not compatible with the [JSON-RPC 2.0 specification](https://www.jsonrpc.org/specification).
+
+**Pros:**
+
+- Simple
+- Well-known and well-understood by users and integrators
+- Integrates reasonably well with web infrastructure (proxies, API gateways, service meshes, caches, etc)
+
+**Cons:**
+
+- No schema support
+- RPC clients must be hand-written
+- Streaming not built into protocol
+- Underspecified types (e.g. numbers and timestamps)
+- Tendermint has its own implementation (not standards compliant, maintenance overhead)
+  - High maintenance cost associated to this
+- Stdlib `jsonrpc` package only supports JSON-RPC 1.0, no dominant package for JSON-RPC 2.0
+- Tooling around documentation/specification (e.g. Swagger) could be better
+- JSON data is larger (offset by HTTP compression)
+- Serializing is slower ([~100% marshal, ~400% unmarshal](https://github.com/alecthomas/go_serialization_benchmarks)); insignificant in absolute terms
+- Specification was last updated in 2013 and is way behind Swagger/OpenAPI
+
+### gRPC + gRPC-gateway (REST + Swagger):
+
+gRPC is a high performant RPC framework. It has been battle tested by a large number of users and is heavily relied on and maintained by countless large corporations.
+
+**Pros:**
+
+- Efficient data retrieval for users, lite clients and other protocols
+- Easily implemented in supported languages (Go, Dart, JS, TS, rust, Elixir, Haskell, ...)
+- Defined schema with richer type system (Protocol Buffers)
+- Can use common schemas and types across all protocols and data stores (RPC, ABCI, blocks, etc)
+- Established conventions for forwards- and backwards-compatibility
+- Bi-directional streaming
+- Servers and clients are be autogenerated in many languages (e.g. Tendermint-rs)
+- Auto-generated swagger documentation for REST API
+- Backwards and forwards compatibility guarantees enforced at the protocol level.
+
+**Cons:**
+
+- Complex system involving cross-language schemas, code generation, and custom protocols
+- Harder to integrate with web infrastructure (proxies, API gateways, service meshes, caches, etc)
+- Type system does not always map cleanly to native language type system; integration woes
+- Many common types require Protobuf plugins (e.g. timestamps and duration)
+- Generated code may be non-idiomatic and hard to use
+- Migration will be disruptive and laborious
+
+There is an option of continuing with amino in the RPC layer. This approach would be best suited for the path of doing the release with the current JSON-RPC & implement gRPC throughout the lifecycle of 0.34.x. This is not preferable as it would mean keeping amino alive in the codebase. 
```

Comment:
But it has the big benefit that it would give people to migrate, right? I think there is a great variety of current users of the current rpc endpoints (not only the sdk).

Reply:
Yes there would be time to migrate. If we keep the current schedule for major releases it looks like it will 6 months at least till its removed